python/cuml/experimental/explainer/permutation_shap.pyx

```diff
@@ -0,0 +1,426 @@
+            row_major = order == "C"
+
+            if dtype == cp.float32:
+                permutation_dataset(handle_[0],
+                                 <float*> masked_ptr,
+                                 <float*> bg_ptr,
+                                 <int> self.n_rows,
+                                 <int> n_cols,
+                                 <float*> row_ptr,
+                                 <int*> idx_ptr,
+                                 <bool> row_major)
+            else:
+                permutation_dataset(handle_[0],
+                                 <double*> masked_ptr,
+                                 <double*> bg_ptr,
+                                 <int> self.n_rows,
+                                 <int> n_cols,
+                                 <double*> row_ptr,
+                                 <int*> idx_ptr,
+                                 <bool> row_major)
+
+            self.handle.sync()
+
+            results = self.model(masked_inputs)
+
+            self.obs_output = results[len(results) / 2]
+
+            results = results.reshape(2 * n_cols + 1, len(self.masker))
+
+            averaged_outs = cp.mean(cp.asarray(self.link(results)), axis=1)
+
+            tpb = 256
+            blks = int((len(idx) + tpb - 1) / tpb)
+
+            if dtype == cp.float32:
+                _row_values_kernel_f32((blks,), (tpb,), (row_values,
+                                                         averaged_outs,
+                                                         idx,
+                                                         len(idx)))
+            else:
+                _row_values_kernel_f64((blks,), (tpb,), (row_values,
+                                                         averaged_outs,
+                                                         idx,
+                                                         len(idx)))
+
+            print("row_values")
+            print(row_values)
+
+            cp.cuda.Stream.null.synchronize()
+
+        row_values /= (2 * npermutations)
+        self.expected_value = averaged_outs[0]
+
+        diff = cp.sum(row_values) - (self.obs_output - self.expected_value)
+
+        if main_effects:
+            del masked_inputs
+            main_effect_values = self.main_effects(row)
+        else:
+            main_effect_values = None
+
+        return row_values
+
+    def main_effects(self, row, inds=None):
+        if inds is None:
+            inds = cp.arange(len(self.masker), dtype=np.float32)
+
+        masked_inputs = cp.empty(
+            shape=((self.n_rows * self.n_cols + self.n_rows), self.n_cols),
+            dtype=self.dtype,
+            order=self.masker.order
+        )
+
+        cdef handle_t* handle_ = \
+            <handle_t*><size_t>self.handle.getHandle()
+        cdef uintptr_t row_ptr, bg_ptr, idx_ptr, masked_ptr
+
+        masked_ptr = masked_inputs.__cuda_array_interface__['data'][0]
```

Comment:
Maybe we should have some accessors for CAI pointers? Seems like a common operation and always looks a little funny